### ISSUE TITLE
Import Specifiers: slightly inconsistent between Acorn and Esprima

### ISSUE THREAD

**Issue by impinball** [estree/estree#11]
Currently, here's the output of each import declaration type (in pseudo-CSON for ease of typing). I also have relevant bugs within this repo about each, if I can find one.

Default import (`import foo from 'bar'`):

```coffee
# Acorn
type: 'ImportSpecifier'
id: Identifier('foo')
name: null
default: true

# Esprima
type: 'ImportDefaultSpecifier'
id: Identifier('foo')

# SpiderMonkey
type: 'ImportSpecifier'
id: Identifier('default')
name: Identifier('foo')
```

Named import (`import {foo} from 'bar'`):

```coffee
# Acorn
type: 'ImportSpecifier'
id: Identifier('foo')
name: null
default: false

# Esprima
type: 'ImportSpecifier'
id: Identifier('foo')
name: null

# SpiderMonkey
type: 'ImportSpecifier'
id: Identifier('foo')
name: Identifier('foo')
```

Namespace import (`import * as foo from 'bar'`):
Bugs: #9

```coffee
# Acorn
type: 'ImportBatchSpecifier'
name: Identifier('foo')

# Esprima
type: 'ImportNamespaceSpecifier'
id: Identifier('foo')

# SpiderMonkey
# (unsupported)
```

Aliased import (`import {foo as bar} from 'bar'`):

```coffee
# Acorn
type: 'ImportSpecifier'
id: Identifier('foo')
name: Identifier('bar')
default: false

# Esprima
type: 'ImportSpecifier'
id: Identifier('foo')
name: Identifier('bar')

# SpiderMonkey
type: 'ImportSpecifier'
id: Identifier('foo')
name: Identifier('bar')
```

I do wonder if the following would be better, a common interface for all of the import specifier types. There are a few backwards-compatibility breaks, but each of them are in areas where Acorn and Esprima already deviate. They are also noted explicitly inline. Any others are truthy/falsy compatible (just as Acorn and Esprima generally are).

```idl
// This is a virtual superclass, purely for clarification.
// No .default field, since it's generally redundant.
interface ImportSpecifierNode <| Node {
  type: string;
  id: Identifier;
  name: Identifier;
}

interface ImportDefaultSpecifier <| ImportSpecifierNode {
  type: "ImportDefaultSpecifier"; // Easier to test, consistent with Esprima
  id: Identifier;
  name: null;
}

interface ImportSpecifier <| ImportSpecifierNode {
  type: "ImportSpecifier";
  id: Identifier;
  name: null; // null if and only if it's not aliased.
}

// This can also be ImportBatchSpecifier -- see my comment on issue #9 in this
// repo.
interface ImportNamespaceSpecifier <| ImportSpecifierNode {
  type: "ImportNamespaceSpecifier";
  id: Identifier; // Esprima has better consistency here.
  name: null;
}
```

This would make for the following examples (using above pseudo-CSON):

```coffee
# import foo from 'bar';
type: 'ImportDefaultSpecifier'
id: Identifier('foo')
name: null

# import {foo} from 'bar';
type: 'ImportSpecifier'
id: Identifier('foo')
name: null

# import {foo as bar} from 'bar';
type: 'ImportSpecifier'
id: Identifier('foo')
name: Identifier('bar')

# import * as foo from 'bar';
type: 'ImportNamespaceSpecifier'
id: Identifier('foo')
name: null
```

WDYT?

**Comment by sebmck** [comment]
Speaking from experience `ImportDefaultSpecifier` is annoying to work with as both of these are parsed differently:

```javascript
import foo from "foo";
import foo as default from "foo";
```

**Comment by RReverser** [comment]
@impinball I've updated examples with SpiderMonkey `Reflect.parse` results as they are pretty close to your original proposal.

**Comment by RReverser** [comment]
I see at least following consistency issues that are still present in current proposal:

```coffee
# import foo from 'bar';
type: 'ImportDefaultSpecifier'
id: Identifier('foo') # local binding in `id`
name: null

# import {foo} from 'bar';
type: 'ImportSpecifier'
id: Identifier('foo') # local binding (export name?) in `id`
name: null

# import {foo as bar} from 'bar';
type: 'ImportSpecifier'
id: Identifier('foo') # now export name in `id`
name: Identifier('bar') # and local binding in `name`

# import * as foo from 'bar';
type: 'ImportNamespaceSpecifier'
id: Identifier('foo') # local binding in `id` again
name: null
```

Also related to #12 - let's figured out once and for all which property we want to use for local binding and which for export name so that there won't be inconsistency between different types of nodes.

**Comment by impinball** [comment]
I was also thinking about instead having both .id and .name returning the
same thing for the unaliased version, with such a check remaining to be an
equality check. If they all should be that way, in that for all of them,
.id should be the identifier and .name should be the local binding, I could
definitely see the benefits in it. The type only denotes special cases of a
single thing.

**Comment by caridy** [comment]
yeah, I think we are all in agreement that this has to be adjusted, and breaking changes everywhere will have to be corrected :)

As I tried to explain in https://github.com/estree/estree/issues/12, and we have two options:

1. `.id` should be the identifier and `.name` should be the local binding
1. or `.id` should be the identifier for the local binding and `name` should be the named export. 

In esprima, I was using 2, but I'm ok with 1 if that makes more sense, specially considering that functions, and other structures are using `name` to denote the local binding.

In specific tools, like es6-module-transpiler, I have seen stuff like `from` to denote the named export from there a specifier is coming from, which is also confusing, but my point is that we don't have to use `id` and `name, we might want to replace one of them with something more meaningful to avoid confusion.

**Comment by impinball** [comment]
I do think that in some cases, particularly in this case, some renaming might be nice, as long as we aren't going the way of Java with `enormouslyLongVariableOrPropertyNames`. ;)

In all seriousness, my main ideas were mostly based on the current state, as it's already what I'm used to. This bug was mostly filed as a sort-of meta-bug to bring up a few specific set of inconsistencies.

**Comment by impinball** [comment]
Would something like this work better?

```coffee
# import foo from 'bar';
type: 'ImportDefaultSpecifier'
import: Identifier('foo')
binding: Identifier('foo')

# import {foo} from 'bar';
type: 'ImportSpecifier'
import: Identifier('foo')
binding: Identifier('foo')

# import {foo as bar} from 'bar';
type: 'ImportSpecifier'
import: Identifier('foo')
binding: Identifier('bar')

# import * as foo from 'bar';
type: 'ImportNamespaceSpecifier'
import: Identifier('foo')
binding: Identifier('foo')
```

In this case, all you really need is there, in an extremely consistent format. Getting all the imported variable names with metadata on each instance is close to trivial (assuming imports are exclusively in the top level, which is mandated by ES6 spec):

```js
function each(xs, f) {
    for (var i = 0, len = xs.length; i < len; i++) {
        f(xs[i]);
    }
}

function Import(binding, name, source) {
  this.binding = binding; // binding name
  this.import = name; // imported name
  this.source = source; // source module
}

function getImportedBindings(ast) {
    var list = [];
    
    each(ast.body, function (node) {
        var node = ast.body[i];
        var type = node.type;
        if (type === 'ImportSpecifier' ||
                type === 'ImportDefaultSpecifier' ||
                type === 'ImportNamespaceSpecifier') {
            each(node.specifiers, function (specifier) {
                list.push(new Import(
                    specifier.binding.name,
                    specifier.import.name,
                    node.source.value));
            });
        }
    });
    
    return list;
}

var bindings = getImportedBindings(acorn.parse(source));
```

I'll also take a look at the export format, but it looks to me as if it requires a bit of further parsing to get it this simple, given its inconsistent format (which isn't too hard to fix).

**Comment by RReverser** [comment]
> but my point is that we don't have to use `id` and `name`, we might want to replace one of them with something more meaningful to avoid confusion

I like this idea the most so far, having both `id` and `name` in one node is pretty confusing.

And yes, @impinball, `import` and `binding` sounds like good options for those.

However, `import` values for special cases don't seem to be correct if we wish those to always point to ids being imported from module:

```javascript
# import foo from 'bar';
type: 'ImportDefaultSpecifier'
import: Identifier('foo') # should be either Identifier('default') or null here
binding: Identifier('foo')

...

# import * as foo from 'bar';
type: 'ImportNamespaceSpecifier'
import: Identifier('foo') # should be null here
binding: Identifier('foo')
```

**Comment by impinball** [comment]
I see what you're saying. And yes, that makes a lot more sense.

```coffee
# import foo from 'bar';
type: 'ImportDefaultSpecifier'
import: Identifier('default')
binding: Identifier('foo')

...

# import * as foo from 'bar';
type: 'ImportNamespaceSpecifier'
import: null
binding: Identifier('foo')
```

**Comment by caridy** [comment]
cool, this is good progress. I think having `null` values does not make sense for something like `ImportNamespaceSpecifier`, because it will never be used.

also, I think this affects export declarations, let's clarify those too because they happen to be symmetric.

**Comment by RReverser** [comment]
@caridy Of course, those might be just missing. Then in base `ImportSpecifierNode` interface only `binding` will be required, and `import` property will exist only in `ImportSpecifier`.

Agree about export declarations - would be nice to have such an explicit property names there as well.

**Comment by impinball** [comment]
@RReverser @caridy 
That could also work. In `ImportDefaultSpecifier` and `ImportNamespaceSpecifier`, the `.import` property is largely pointless. Also, it would provide guaranteed differentiation between `import foo from 'bar'` and `import {default as foo} from 'bar'`, which is useful for syntax transformers that aren't minifiers.

Would this potentially have to include limited parsing of the values themselves, such as in `export function foo() {}`? I expect it probably would, but iff it is a declaration.

Here's a little example from current output. Both Esprima and Acorn emit this:

```coffee
# export {foo};
type: 'ExportSpecifier'
id: Identifier('foo')
name: null

# export {foo as bar};
type: 'ExportSpecifier'
id: Identifier('foo')
name: Identifier('bar')

# export let foo = 1;
type: 'ExportDeclaration'
default: false
declaration:
    type: 'VariableDeclaration'
    declarations: [
        type: 'VariableDeclarator'
        id: Identifier('foo')
        init: Literal(1)
    ]
    kind: 'let'
    specifiers: []
    source: null
```

The latter is a little bit excessive IMO, although it would probably complicate a few things if the `declaration` part was alone fixed. In this case, the best way to fix that specific case would probably be conceptually this:

```js
node.specifiers = node.declaration.declarations.map(decl => ({
    type: 'ExportSpecifier',
    id: decl.id,
    name: decl.id,
}));
```

Part of why I'm not wanting to remove `.declaration` from that is that visitors may be interested in using/renaming those nodes without regards to if it's being exported. A great example would be a transformer for React users, where the class name still needs registered, and there is no guarantee it'll be correct with name mangling, much less debugging.

```js
export default class MyComponent extends React.Class {
    render() {
        // blah...
    }
}

// Needs transpiled to:

export default var MyComponent = React.createClass('MyComponent', {
    render: function () {
        // blah...
    }
});
```

For export specifiers themselves, maybe this might work?

```coffee
# export {foo};
type: 'ExportSpecifier'
export: Identifier('foo')
local: Identifier('foo')

# export {foo as bar};
type: 'ExportSpecifier'
export: Identifier('foo')
local: Identifier('bar')

# export let foo = 1;
# In addition to the standard ExportSpecifier.declaration
type: 'ExportSpecifier'
export: Identifier('foo')
local: Identifier('foo')

# export function foo() {}
# In addition to the standard ExportSpecifier.declaration
type: 'ExportSpecifier'
export: Identifier('foo')
local: Identifier('foo')

# export default foo;
type: 'ExportDefaultSpecifier'
export: Identifier('default')
local: null

# export default function foo() {}
type: 'ExportDefaultSpecifier'
export: Identifier('default')
local: Identifier('foo')

# export default let foo = 1;
type: 'ExportDefaultSpecifier'
export: Identifier('default')
local: Identifier('foo')

# export {default as foo};
type: 'ExportSpecifier'
export: Identifier('default')
local: Identifier('foo')

# export {foo} from 'bar';
type: 'ExportSpecifier'
export: Identifier('foo')
local: Identifier('foo')

# export * from 'bar';
type: 'ExportBatchSpecifier'
# Everything else here is irrelevant
```

Or, to summarize the above into a little schema:

```idl
interface ExportSpecifier {
    type: "ExportSpecifier";
    export: Identifier;
    local: Identifier;
}

interface ExportDefaultSpecifier {
    type: "ExportDefaultSpecifier";
    export: Identifier;
    local: Identifier or null;
}

interface ExportBatchSpecifier {
    type: "ExportBatchSpecifier";
}
```

`ExportSpecifier` here is relatively similar to `ImportSpecifier`, and `ExportDefaultSpecifier` differs in that the exported binding can be easily relevant to other applications.

**Comment by mikesherov** [comment]
Would love some feedback from @ariya or @jeffmo here. And @caridy please don't merge your espree or until we can get confirm from Esprima. No reason to create more divergence.

**Comment by caridy** [comment]
guys, does anyone disagree with this proposal? can we move forward with it? if not, we just need a volunteer to formalize it.

**Comment by RReverser** [comment]
I'm fine with import representations above, but I remember I still had some concerns with export, need to look again and will respond here. Maybe better to split and formalize at least import for now?

**Comment by jeffmo** [comment]
Reading @impinball 's last post as a summary, a question:

```
# export {foo as bar};
type: 'ExportSpecifier'
export: Identifier('foo')
local: Identifier('bar')
```

I think just a typo, but to clarify: Wouldn't the "local" be `foo` and the "export" be `bar` here?

Otherwise I think this looks pretty great. Two minor nits:

(1) can we call the `export` key on the Export* nodes `exported` (just to make it fit the vocalizations: "the *exported* binding" and "the *local* binding name"
(2) On the Import* nodes, can we rename the `import` key -> `imported` and the `binding` key -> `local (to fit "the *imported* binding" and "the *local* binding" and to match up `local` with the Export* nodes)

**Comment by jeffmo** [comment]
Also it could be nice to look at a complete, flattened spec resulting from the iterations in this thread before committing.

I guess maybe that can just go in the PR, though?

**Comment by caridy** [comment]
@jeffmo yes, that should be a typo, and I think your interpretation is correct.

as for the flattened spec, that discussion can happen in the actual PR.

**Comment by mikesherov** [comment]
We'll also be discussing this at next Esprima meeting, Wednesday at 12PST, in #esprima-meeting on freenode. We'll achieve resolution there.

Sorry for the delay, but group consensus amongst our team is primary concern before we move forward with this one!

Thanks for patience as always. @caridy, can you attend?

**Comment by caridy** [comment]
@mikesherov will try to attend, but I'm on the road next week.

**Comment by ariya** [comment]
This looks good. And I agree with @jeffmo, it's easier to evaluate once the corresponding PR is available.

**Comment by impinball** [comment]
@jeffmo 

1. Good catch on the typo.
2. The renaming makes sense. I'll make the PR shortly. (I finally have some time to write one, too.)

**Comment by impinball** [comment]
@all Here's my PR for imports. (#35) I'll get to exports soon.

**Comment by mikesherov** [comment]
Now just waiting for #38 to land before we can close this...

**Comment by mikesherov** [comment]
All done! Thanks @all!
